File: src/lab02/libs/tuples.py
class Tuple:
    def __init__(self, fio, group, gpa):
        # exception
        if not isinstance(fio, str) or not isinstance(group, str) or not isinstance(gpa, (int, float)) or gpa < 0 or gpa > 5 or len(fio) == 0 or len(group) == 0:
            raise TypeError("Failed to construct tuple. Wrong data")
        
        # fio
        self.fio = list(dict.fromkeys(fio.split(' ')))
        if '' in self.fio:
            self.fio.remove('')
        self.fio = [name.capitalize() for name in self.fio]
        
        # group
        self.group = group.replace(" ", "")
        
        # gpa
        self.gpa = format(gpa, '.3g')

    def format_record(self):
        second_name = self.fio[0]
        first_name_short = self.fio[1][0]
        if len(self.fio) == 3:
            third_name_short = self.fio[2][0]
            a = f"{second_name} {first_name_short}. {third_name_short}."
        else:
            a = f"{second_name} {first_name_short}."
        return f"{a}, гр. {self.group}, GPA {self.gpa}"
    

def format_record(rec: tuple[str, str, float]) -> str:
    fio = list(dict.fromkeys(rec[0].split(' ')))
    if '' in fio:
        fio.remove('')
    fio = [name.capitalize() for name in fio]
    
    # group
    group = rec[1].replace(" ", "")
    
    # gpa
    gpa = format(rec[2], '.3g')

    second_name = fio[0]
    first_name_short = fio[1][0]
    if len(fio) == 3:
        third_name_short = fio[2][0]
        a = f"{second_name} {first_name_short}. {third_name_short}."
    else:
        a = f"{second_name} {first_name_short}."
    return f"{a}, гр. {group}, GPA {gpa}"

File: src/lab02/libs/tuples.js
export class Tuple {
    constructor(fio, group, gpa)
    {
        //exception
        if (typeof fio !== "string" || typeof group !== "string" || typeof gpa !== "number" || gpa < 0 || gpa > 5 || fio.length === 0 || group.length === 0) throw new TypeError("Failed to construct tuple. Wrong data");
        //code

        //fio
        this.fio = Array.from(new Set(fio.split(' ')));
        if (this.fio.indexOf('') !== -1)
            this.fio.splice(this.fio.indexOf(''), 1);
        for (var i = 0; i < this.fio.length; i++){
            var ch = this.fio[i][0].toUpperCase();
            this.fio[i] = ch + this.fio[i].slice(1, this.fio[i].length);
        }
        //group
        this.group = group;
        while (group.includes(" "))
        {
            group.delete(" ");
        }
        //gpa
        this.gpa = gpa.toPrecision(3);
        Object.freeze(this);
    }
    /**
     * 
     */
    format_record() {
        var secondName = this.fio[0];
        var firstNameShort = this.fio[1][0];
        if (this.fio.length === 3){
            var thirdNameShort = this.fio[2][0];
            var a = secondName + " " + firstNameShort + ". " + thirdNameShort + ".";
        }
        else
            var a = secondName + " " + firstNameShort + ".";
        return `${a}, гр. ${this.group}, GPA ${this.gpa}`;
    }


}

export function format_record(array)
{
    if (!Array.isArray(array) || !array.length === 3) throw new TypeError("Wrong data!");
    return  new Tuple(array[0], array[1], array[2]).format_record()

}
